spec/smart_app_launch/client_assertion_builder_spec.rb

```diff
@@ -7,21 +8,52 @@
   let(:aud) { 'AUD' }
   let(:jwks) { SMARTAppLaunch::JWKS.jwks }
 
+  def build_and_decode_jwt(encryption_method, kid)
+    jwt = described_class.build(client_auth_encryption_method: encryption_method, iss:, sub:, aud:, kid: kid)
+    return JWT.decode(jwt, kid, true, algorithms: [encryption_method], jwks:)
+  end
+
   describe '.build' do
-    it 'creates a valid JWT' do
-      client_auth_encryption_methods.each do |client_auth_encryption_method|
-        jwt = described_class.build(client_auth_encryption_method:, iss:, sub:, aud:)
+    context 'with unspecified key id' do
+      it 'creates a valid JWT' do
+        client_auth_encryption_methods.each do |client_auth_encryption_method|
+          payload, header = build_and_decode_jwt(client_auth_encryption_method, nil)
+
+          expect(header['alg']).to eq(client_auth_encryption_method)
+          expect(header['typ']).to eq('JWT')
+          expect(payload['iss']).to eq(iss)
+          expect(payload['sub']).to eq(sub)
+          expect(payload['aud']).to eq(aud)
+          expect(payload['exp']).to be_present
+          expect(payload['jti']).to be_present
+        end
+      end
+    end
 
-        payload, header = JWT.decode(jwt, nil, true, algorithms: [client_auth_encryption_method], jwks:)
+    context 'with specified key id' do
+      it 'creates a valid JWT with correct algorithm and kid' do
+        encryption_method = 'ES384'
+        kid = '4b49a739d1eb115b3225f4cf9beb6d1b'
+        payload, header = build_and_decode_jwt(encryption_method, kid)
 
-        expect(header['alg']).to eq(client_auth_encryption_method)
+        expect(header['alg']).to eq(encryption_method)
         expect(header['typ']).to eq('JWT')
+        expect(header['kid']).to eq(kid)
         expect(payload['iss']).to eq(iss)
         expect(payload['sub']).to eq(sub)
         expect(payload['aud']).to eq(aud)
         expect(payload['exp']).to be_present
         expect(payload['jti']).to be_present
       end
+
+      it 'throws exception when key id not found for algorithm' do
+        encryption_method = 'RS384'
+        kid = '4b49a739d1eb115b3225f4cf9beb6d1b'
+
+        expect {
+          build_and_decode_jwt(encryption_method, kid)
+        }.to raise_error(RuntimeError)
```

Comment:
So this error would be the result of bad inputs? If so, it should be an `Inferno::Exceptions::AssertionException`. If this would be caused by `private_key` being `nil`, the intention would be clearer by having a guard against it being `nil` that raises this exception rather than rescuing `NoMethodError`.

Reply:
Thanks!  Yes, this is a result of bad inputs.  I've updated this to check for a nil `private_key` and to raise `Inferno::Exceptions::AssertionException` instead.